requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -0,0 +1,600 @@
+
+```json
+{
+  "insight": "FUNNELS",
+  "date_from": "-14d",
+  "actions": [
+    {
+      "id": "2",
+      "type": "actions",
+      "order": 2,
+      "name": "definitely not an action"
+    }
+  ],
+  "events": [
+    {
+      "id": "$pageview",
+      "name": "$pageview",
+      "type": "events",
+      "order": 0,
+      "custom_name": "Visited a page"
+    },
+    {
+      "id": "viewed dashboard",
+      "type": "events",
+      "order": 1,
+      "name": "viewed dashboard",
+      "properties": [
+        {
+          "key": "created_at",
+          "value": "2022-11-14",
+          "operator": "is_date_exact",
+          "type": "group",
+          "group_type_index": 1
+        },
+        {
+          "key": "$browser",
+          "value": ["Chrome"],
+          "operator": "exact",
+          "type": "event"
+        },
+        {
+          "key": "email_opt_in",
+          "value": ["true"],
+          "operator": "exact",
+          "type": "person"
+        }
+      ]
+    }
+  ],
+  "display": "FunnelViz",
+  "new_entity": [],
+  "interval": "day",
+  "properties": [],
+  "funnel_viz_type": "steps",
+  "exclusions": []
+}
+```
+
+If I want to open "step 2 dropoff", the filter just gets these extra fields, and gets sent to a different API endpoint (`/insights/funnel` vs `/person/funnel`):
+
+```json
+{
+  "funnel_step_breakdown": [""],
+  "funnel_step": -2
+}
+```
+
+Here's a path with a global filter:
+
+```json
+{
+  "insight": "PATHS",
+  "properties": {
+    "type": "AND",
+    "values": [
+      {
+        "type": "AND",
+        "values": [
+          {
+            "key": "$anon_distinct_id",
+            "value": "e",
+            "operator": "not_icontains",
+            "type": "event"
+          }
+        ]
+      }
+    ]
+  },
+  "end_point": "http://amazingstore.com/cart",
+  "step_limit": 8,
+  "include_event_types": ["$pageview"],
+  "path_groupings": [],
+  "exclude_events": [],
+  "date_from": "-14d",
+  "date_to": null,
+  "funnel_filter": {
+    "date_from": "-14d"
+  },
+  "local_path_cleaning_filters": []
+}
+```
+
+Moving away from insights, here's a query under cohorts for "match persons who match any criteria: completed an event multiple times. $autocapture exactly 5 times in the last 30 days"
+
+```json
+{
+  "properties": {
+    "type": "OR",
+    "values": [
+      {
+        "type": "OR",
+        "values": [
+          {
+            "key": "$autocapture",
+            "type": "behavioral",
+            "value": "performed_event_multiple",
+            "negation": false,
+            "operator": "exact",
+            "event_type": "events",
+            "time_value": "30",
+            "time_interval": "day",
+            "operator_value": 5
+          }
+        ]
+      }
+    ]
+  }
+}
+```
+
+## Proposal: nestable typed filter objects
+
+Instead of a bespoke and scary monolithic filter object, I propose to split this up into **nestable typed filter objects**.
```

Comment:
This is much less work than we think it is, and I don't think we need to start from scratch. I think we can use [these serializers](https://github.com/PostHog/posthog/blob/master/posthog/api/insight_serializers.py#L143) that I created for API documenting, and just create one for each insight type that we're proposing, plus a handful more shapes for the new type of queries.

The added benefit is that we can strongly type this + automatically create an OpenAPI library so querying in the frontend becomes trivial.